File: 백준/Gold/7576. 토마토/토마토.py
import sys
from collections import deque


def initialize_graph(n):
    graph = {}
    for i in range(n):
        graph[i] = list(map(int, sys.stdin.readline().strip().split()))

    return graph


# 모든 토마토가 익었다면 True
def check_graph(n, m, graph):
    for i in range(n):
        for j in range(m):
            if graph[i][j] == 0:
                return False
    return True


def find_start_point(n, m, graph):
    start_points = []
    for i in range(n):
        for j in range(m):
            if graph[i][j] == 1:
                start_points.append([i, j])
    return start_points


if __name__ == "__main__":
    m, n = map(int, sys.stdin.readline().strip().split())
    graph = initialize_graph(n)

    days = 0
    start_points = find_start_point(n, m, graph)

    dir_x = [1, 0, -1, 0]
    dir_y = [0, 1, 0, -1]

    visited = [[False] * m for i in range(n)]

    while start_points:
        queue = deque()
        tmp_start_points = []
        for start_point in start_points:
            queue.append(start_point)
            visited[start_point[0]][start_point[1]] = True

        while queue:
            cur = queue.popleft()
            cur_x = cur[0]
            cur_y = cur[1]

            for i in range(4):
                new_x = cur_x + dir_x[i]
                new_y = cur_y + dir_y[i]

                if 0 <= new_x < n and 0 <= new_y < m and graph[new_x][new_y] == 0 and not visited[new_x][new_y]:
                    visited[new_x][new_y] = True
                    graph[new_x][new_y] = 1
                    tmp_start_points.append([new_x, new_y])
        if len(tmp_start_points) != 0:
            days += 1
        start_points = tmp_start_points

    if check_graph(n, m, graph):
        print(days)
    else:
        print(-1)

File: 백준/Gold/7569. 토마토/토마토.java
import java.util.*;
import java.io.*;

class Node {
    int x;
    int y;
    int h;

    public Node(int x, int y, int h) {
        this.x = x;
        this.y = y;
        this.h = h;
    }

    public int getX() {
        return this.x;
    }

    public int getY() {
        return this.y;
    }

    public int getH() {
        return this.h;
    }
}

public class Main {
    public static boolean checkZero(int[][][] graph, int m, int n, int h) {
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    if (graph[i][j][k] == 0) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public static void printGraph(int[][][] graph, int m, int n, int h) {
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                System.out.println(Arrays.toString(graph[i][j]));
            }
            System.out.println("------");
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int m, n, h;
        int days = 0;
        String[] input = br.readLine().split(" ");
        m = Integer.parseInt(input[0]);
        n = Integer.parseInt(input[1]);
        h = Integer.parseInt(input[2]);

        int[][][] graph = new int[h][n][m];
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                int tmpIdx = 0;
                for (String s : br.readLine().strip().split(" ")) {
                    graph[i][j][tmpIdx] = Integer.parseInt(s);
                    tmpIdx++;
                }
            }
        }

        boolean[][][] visited = new boolean[h][n][m];
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    visited[i][j][k] = false;
                }
            }
        }

        Deque<Node> queue = new ArrayDeque<>();
        // 시작 지점 큐에 넣기
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    if (graph[i][j][k] == 1) {
                        queue.add(new Node(j, k, i));
                        visited[i][j][k] = true;
                    }
                }
            }
        }

        int[] dirX = {1, 0, -1, 0};
        int[] dirY = {0, 1, 0, -1};
        int[] dirH = {1, -1};

        while (!queue.isEmpty()) {
            days++;
            Deque<Node> tmpQueue = new ArrayDeque<>();
            while (!queue.isEmpty()) {
                tmpQueue.push(queue.pollFirst());
            }

            while (!tmpQueue.isEmpty()) {
                Node curNode = tmpQueue.poll();
                int curX = curNode.getX();
                int curY = curNode.getY();
                int curH = curNode.getH();

                // 현재 레벨에서 앞뒤좌우
                for (int di = 0; di < 4; di++) {
                    int newX = curX + dirX[di];
                    int newY = curY + dirY[di];
                    if (newX >= 0 && newX < n && newY >= 0 && newY < m) {
                        if (graph[curH][newX][newY] == 0 && !visited[curH][newX][newY]) {
                            visited[curH][newX][newY] = true;
                            graph[curH][newX][newY] = 1;
                            queue.push(new Node(newX, newY, curH));
                        }
                    }
                }

                // 높이 위 아래
                for (int dh : dirH) {
                    int newH = curH + dh;
                    if (newH >= 0 && newH < h) {
                        if (graph[newH][curX][curY] == 0 && !visited[newH][curX][curY]) {
                            visited[newH][curX][curY] = true;
                            graph[newH][curX][curY] = 1;
                            queue.push(new Node(curX, curY, newH));
                        }
                    }
                }
            }
        }

        if (checkZero(graph, m, n, h)) {
            System.out.println(days - 1);
        } else {
            System.out.println(-1);
        }
    }
}
